Framework/API/src/FunctionProperty.cpp

```diff
@@ -16,20 +16,24 @@ namespace API {
  *  Sets the property names but initialises the function pointer to null.
  *  @param name :: The name to assign to the property
  *  @param direction :: The direction of the function (i.e. input or output)
+ *  @param optional :: A flag indicating whether the property is optional or mandatory.
  */
-FunctionProperty::FunctionProperty(const std::string &name, const unsigned int direction)
+FunctionProperty::FunctionProperty(const std::string &name, const unsigned int direction,
+                                   const PropertyMode::Type optional)
     : Kernel::PropertyWithValue<std::shared_ptr<IFunction>>(
-          name, std::shared_ptr<IFunction>(), Kernel::IValidator_sptr(new Kernel::NullValidator()), direction) {}
+          name, std::shared_ptr<IFunction>(), Kernel::IValidator_sptr(new Kernel::NullValidator()), direction),
```

Comment:
This needs to use make shared instead of new, which should only be used with Qt components:
- New can leak if the object fails to construct
- Make lifetime clear to static analysis tools (whereas new is a "trust me I've got this okay")
- Easier to read

This line can be written as:
```
    : Kernel::PropertyWithValue<std::shared_ptr<IFunction>>(name, std::shared_ptr<IFunction>(),
                                                            std::make_shared<Kernel::NullValidator>(), direction),
```
And C++ is smart enough to see that `Kernel::NullValidator` is a derived type of `IValidator` and will create the right abstract pointer on the object for you